### ISSUE TITLE
Lightmap baking creates noticeable seams on meshes

### ISSUE THREAD

**Issue by reptofrog** [godotengine/godot#90929]
### Tested versions

v4.2.2.stable.official [15073afe3]

### System information

Godot v4.2.2.stable - Fedora Linux 39 (Workstation Edition) - Wayland - Vulkan (Forward+) - integrated AMD Radeon Graphics (RADV RENOIR) () - AMD Ryzen 7 5700G with Radeon Graphics (16 Threads)

### Issue description

Lightmap baking creates noticeable seams on meshes.

The seams appear with and without the denoiser.

![image](https://github.com/godotengine/godot/assets/36744029/9269012b-1e9f-40ec-bc69-82959bc4de20)

### Steps to reproduce

—

### Minimal reproduction project (MRP)

[Lightmap Seams.zip](https://github.com/godotengine/godot/files/15047529/Lightmap.Seams.zip)

**Comment by passivestar** [comment]
Can confirm, in my experience more often than not cylinders bake with stripes

I wasn't able to get rid of them by changing any settings

<img width="1007" alt="image" src="https://github.com/godotengine/godot/assets/60579014/c6b0847c-aaa7-4b87-806b-3fa07214baeb">

**Comment by patwork** [comment]
Yep, there are some serious problems with lightmaps even in 4.3 master.

scene:
![ll](https://github.com/godotengine/godot/assets/224215/db6558b7-4577-4aed-8b3c-caa964634831)

lightmap:
![ll2](https://github.com/godotengine/godot/assets/224215/7a96d4f2-9ccf-461b-9e59-3b59c45ab94f)

https://github.com/godotengine/godot/issues/87815

**Comment by passivestar** [comment]
> Can confirm, in my experience more often than not cylinders bake with stripes

This is what they look like in the lightmap:

![stripes](https://github.com/godotengine/godot/assets/60579014/072c5382-661b-4e0d-a00b-7c2255f9ddc8)

**Comment by lander-vr** [comment]
Aside from Passivestar's issue, which I don't think is related to this issue post and probably needs its own issue report, the these baking errors come down to poor lightmapper setups, which I think can be attributed to a lack of documentation on how to use a lightmapper. If I can find the time, I could work on a guide to add to the docs going over how to achieve clean bakes with LightmapGI, and covering some common pitfalls that arise.

The issues shown here come from a combination of a too low lightmap resolution (which could be attributed to Godots very low default texel density) and a bake that's not converged well enough (quality is set too low).
This results in a very low resolution, very noisy render, which is practically impossible to denoise to a point where you get a qualitative result.
You can see in this first example bake (OPs settings) how practically every neighboring pixel has a completely different color:
![image](https://github.com/godotengine/godot/assets/31851431/2d80af90-7ad9-41dd-8220-e0a1baf99fac)
With this as a starting point, it is expected to get a poor end result even using denoising.

First thing I did was to increase bounces, this gives the rays more opportunities to bounce off of surfaces and hit the backs of the cylinders/sphere:
![image](https://github.com/godotengine/godot/assets/31851431/b783819c-62ce-4d3d-af69-01b645aaeffe)


Increasing the quality to ultra, without denoising, you can see how the overall result is much cleaner. This will give much better results from the denoiser, because instead of having to go from pure noise to a clean smooth result, it only has to work away some smaller inconsistencies. The seams are still very visible though:
![image](https://github.com/godotengine/godot/assets/31851431/dd7b3d1c-3ad6-420e-b583-4e1a2388e773)

To reduce the seams, we need to increase how much detail is shown, which we do by increasing the texel density. You can see the seams practically all disappear (except the sphere but I'll get to that):
![image](https://github.com/godotengine/godot/assets/31851431/aac25fb7-c150-42d7-a9ff-08a46534a3ec)

Adding the denoiser to this gives us a qualitative clean result with no seams:
![image](https://github.com/godotengine/godot/assets/31851431/2015c549-3964-4067-8a1c-a915c1c31636)

I do think it is strange that those leaks in the corners still occur though. Luckily the denoiser is able to smooth those away, but it is reducing AO that would and should be present there. I think https://github.com/godotengine/godot/pull/81872 should've resolved that. Cc: @DarioSamo

The sphere still shows a bad seam, and this comes down to the UV2 unwrap:
![image](https://github.com/godotengine/godot/assets/31851431/4f03999a-0562-43e6-ac83-6598db3c2c0f)

Spheres would be more appropriately unwrapped as two individual hemispheres.

As for @passivestar's issue:
> Can confirm, in my experience more often than not cylinders bake with stripes
> 
> I wasn't able to get rid of them by changing any settings
> <img alt="image" width="1007" src="https://private-user-images.githubusercontent.com/60579014/324186296-c6b0847c-aaa7-4b87-806b-3fa07214baeb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ5Nzg3NDksIm5iZiI6MTcxNDk3ODQ0OSwicGF0aCI6Ii82MDU3OTAxNC8zMjQxODYyOTYtYzZiMDg0N2MtYWFhNy00Yjg3LTgwNmItM2ZhMDcyMTRiYWViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA2VDA2NTQwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiMzMzZWY4OThkMDY5MTBlOTExYWJlNGEwNDZlYWNjY2E4NGM4ZWE0NDg4ZTcyNTMzOTIxZTcyYWQyMTQxYzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.11akPOxEMbMYF5cCV0MnZheD38P6yN5WX0ri0LCsVCk">

This one is strange, because it looks like those two lines on the right side are showing up in the middle of your lightmap and not at the borders of your UV islands. The two lines on the left side have a hard cutoff caused by the edge of the lightmap UV island, which would indicate an issue with padding or something along those lines, but those two on the right side don't seem to have that so it looks like they're just random rows of pixels in the middle of a UV island that are experiencing light leakage somehow.

**Comment by patwork** [comment]
I am sorry but I cannot agree with this. Of course increasing the resolution of lightmaps will improve their appearance, but the problem of light leakage at the edges remains, it's just a little sharper.

Please, compare the lightmaps from 3.x and 4.x - the ones from 3.x are clean from edge to edge, while the ones from 4.x have strange lighter borders. Denoising is not the solution as it only blurs these borders making them appear even larger at times.

All the time something is wrong with the accuracy of the calculation, perhaps the UV boundary conditions? Where are those black pixels in the corners of the faces in the lightmaps from 4.x coming from? Where do these lighter edges come from?

**Comment by lander-vr** [comment]
This issue isn't about light leaks at corners though, it's about seams appearing on meshes. Those are a universal thing with any lightmapper and the steps I have gone over is how you'd go about reducing those seams in other lightmappers too.

> Please, compare the lightmaps from 3.x and 4.x - the ones from 3.x are clean from edge to edge, while the ones from 4.x have strange lighter borders. Denoising is not the solution as it only blurs these borders making them appear even larger at times.

I addressed those leaks and cc'd Dario because he'll likely have a better idea what's causing those since he worked on a PR to resolve that particular issue. I didn't mean to say denoising is an appropriate solution for this issue, just that in this situation it does not show up very visibly after denoising with the drawback of reducing AO that would be present in those corners.

My guess is that when UV islands are not aligned to the edges of texels, the center of a pixel (due to half-pixel offset) can be in a position that's not on the actual mesh, which would cause it to shoot out a ray that's not originating from a point where the surface of that mesh exists. This means that when you have a floorplane and simple planes as walls around it, that ray can be shot out from behind a wall (because the center of a pixel may be there), thus not intersecting with that wall, in this case returning the environment lighting. This would also explain why those leaks get severely reduced when increasing resolution.

That said, I would suggest to open an issue that specifically addresses those particular leaks, if it doesn't exist already.

> Where are those black pixels in the corners of the faces in the lightmaps from 4.x coming from?

Those single pixels are in the padded areas and don't show up on meshes themselves, it's likely something to do with how the padding algorithm works but I don't really see how those would cause any issues since it's not within the lightmap UVs, except _maybe_ when generating mipmaps.

**Comment by DarioSamo** [comment]
> My guess is that when UV islands are not aligned to the edges of texels, the center of a pixel (due to half-pixel offset) can be in a position that's not on the actual mesh, which would cause it to shoot out a ray that's not originating from a point where the surface of that mesh exists. This means that when you have a floorplane and simple planes as walls around it, that ray can be shot out from behind a wall (because the center of a pixel may be there), thus not intersecting with that wall, in this case returning the environment lighting. This would also explain why those leaks get severely reduced when increasing resolution.

There's a suspicion that is very much the case for simple meshes that weren't imported. Here's the issue where I last suspected some [meshes are being generated with the wrong UV2](https://github.com/godotengine/godot/issues/69126#issuecomment-1734730220).

**Comment by mikatomik** [comment]
Dropping in to say I'm having this issue with non-cylinder meshes as well. I'm on 4.2.2 Windows.

I made a small modular kit, the meshes are manifold and normals are correct.  Each piece of the kit has a crazy seam on it after baking. Maybe it's related? Here's screenshots in lighting view:

Before LightmapGI:
![image](https://github.com/godotengine/godot/assets/109499937/83c6402b-f861-494d-af5f-1ea1dbbe83c4)

After LightmapGI:
![image](https://github.com/godotengine/godot/assets/109499937/6258a85a-eeff-4d5e-a80b-9357190fd62d)

and without Lighting view:
![image](https://github.com/godotengine/godot/assets/109499937/4673b9f0-abdd-4707-b3a0-97cde0d6b509)

Normals in Blender:
![image](https://github.com/godotengine/godot/assets/109499937/987a7240-7a45-4654-8000-bfaf5d7ae12a)

**Comment by lander-vr** [comment]
@mikatomik This is a classic issue with lightmappers and modular kits. Related to #82607, and will be (mostly) resolved by https://github.com/godotengine/godot/pull/91601. @Calinou has made a [proof-of-concept implementation](https://github.com/godotengine/godot/issues/82607#issuecomment-2108917575) for lightmap downsampling, which together with the denoiser-range property will give the tools needed to entirely resolve seams between modular meshes and support modular workflows.

**Comment by mikatomik** [comment]
@lander-vr Ah, thanks for the info. Wasn't aware that's a common issue with modular pieces. Shows what I know about the lower-level stuff. 😅 Hopefully Calinou's work will straighten this out down the line. Thanks!

**Comment by martynBarac** [comment]
<img width="666" height="567" alt="Image" src="https://github.com/user-attachments/assets/14484b68-dd69-48f9-aceb-e503e9672647" />

Issue on cylinders still persists in 4.5. I'm guessing because things are not smoothed where there is a UV seam, but the geometry is still connected.

**Comment by Saul2022** [comment]
> Issue on cylinders still persists in 4.5. I'm guessing because things are not smoothed where there is a UV seam, but the geometry is still connected.

Have you tried with downsampling?

**Comment by martynBarac** [comment]
> > Issue on cylinders still persists in 4.5. I'm guessing because things are not smoothed where there is a UV seam, but the geometry is still connected.
> 
> Have you tried with downsampling?

Somehow turning up supersampling makes it worse. Here is how it looks on 8:

<img width="729" height="618" alt="Image" src="https://github.com/user-attachments/assets/21534d9f-5433-4167-ad98-f10ce6c8942d" />

**Comment by Saul2022** [comment]
> Somehow turning up supersampling makes it worse. Here is how it looks on 

Oh.. i guess you should tell that  in rocket chat since it weird it gets worse with downsampling.  Is the mesh a godot primitive mesh ? Have you tried   adding a cylinder mesh from blender ?

**Comment by martynBarac** [comment]
> Oh.. i guess you should tell that in rocket chat since it weird it gets worse with downsampling. Is the mesh a godot primitive mesh ? Have you tried adding a cylinder mesh from blender ?

Yes it is a primitive mesh. Here it is compared to two blender cylinders with different UV2 maps (Displayed above the cylinders). 

<img width="1154" height="623" alt="Image" src="https://github.com/user-attachments/assets/4c632bb4-3156-42b6-9588-8f7bc66598cd" />

The severity also depends on the angle of the light. The middle cylinder has a manually set UV2 map that keeps all the edges connected. It seems to fix the crease issue but gets a wave pattern instead since the unwrap is quite distorted from the original mesh. 

I think the Lightmapper is not taking into account that the edges are connected in the mesh when they are disconnected in the UV map.

**Comment by Saul2022** [comment]
> I think the Lightmapper is not taking into account that the edges are connected in the mesh when they are disconnected in the UV map.

Seems like so you should report it at godot rendering channel from rocket chat to get more visibility.

**Comment by expenses** [comment]
I did some investigating and created a hacky fix that seems to greatly improve the situation: https://github.com/expenses/godot/commit/437a9f10a14157126f28dfc62766fe81a2e43fe7

It: 
- reverts https://github.com/godotengine/godot/issues/69126 - seems like an existing hack
- turns off the conservative rasterization emulation  - I'm not sure that conservative rasterization is really a good idea with lightmaps. I think making sure that every triangle is represented in UV2 is better.
- Disabled soft shadows where `trace_direct_light` is called - probably because it pushes the ray origin out of the mesh.

I don't think things are perfect but they're significantly better with these changes.
